api/src/main/java/org/apache/iceberg/Schema.java

```diff
@@ -54,10 +54,14 @@
   public Schema(List<NestedField> columns, Map<String, Integer> aliases) {
     this.struct = StructType.of(columns);
     this.aliasToId = aliases != null ? ImmutableBiMap.copyOf(aliases) : null;
+
+    // validate the schema through IndexByName visitor
+    TypeUtil.indexByName(struct);
```

Comment:
As long as `indexByName` is being called, we may as well capture the result and use it for the name to ID map. Can you update these constructors to call `lazyNameToId` instead?

Reply:
Since lazyNameToId has one more check for uniqueness for value, some unit tests failed due to they don't have unique field ID such as `testMixedTypes`.  And I think the lazy initialization is not lazy if we put it in a constructor since `NameToId` is always null when constructing. So can we just call IndexByName in constructor? It would be an easy way.

Reply:
Can we fix those tests by reassigning new IDs to types? I think that would be a better solution.

Reply:
Do you mean update scheme definition in unit tests by reassigning new IDs? For example, there is a constant field `SUPPORTED_PRIMITIVES` in `testMixedTypes` that is used by other nested fields, do we need to remove the constant and use raw definition with new IDs? Or do you mean to use `reassignIds` from TypeUtils?

Reply:
No, I mean to use `TypeUtil.assignFreshIds`.